weighted_graphs.py

```diff
@@ -0,0 +1,214 @@
+
+        if proposed_cost < costs_to_start_vert[neighbour_vertex]:
+        # found a shorter path to this neighbour_vertex! update the dict!
+          print(f"from {starting_vertex} to {neighbour_vertex}, proposed cost {proposed_cost} < current cost {costs_to_start_vert[neighbour_vertex]}. update dict")
+          costs_to_start_vert[neighbour_vertex] = proposed_cost
+
+          # add this to the priority queue
+          heapq.heappush(priority_queue, (proposed_cost, neighbour_vertex))
+
+        else:
+          print(f"from {starting_vertex} to {neighbour_vertex}, proposed cost {proposed_cost} >= current cost {costs_to_start_vert[neighbour_vertex]}. don't update dict")
+
+        print(f"processed neighbour vertex {neighbour_vertex}. costs_to_{starting_vertex}: {costs_to_start_vert})\n")
+
+    print(f"\n**ending**\nfinal costs to each vertex from vertex {starting_vertex}: {costs_to_start_vert}")
+    print(f"original graph: {graph}")
+    return costs_to_start_vert
+
+  def network_delay_time(self, times, N, K):
+    """
+      from https://leetcode.com/problems/network-delay-time/
+      There are N network nodes, labelled 1 to N.
+
+      Given times, a list of travel times as directed edges
+      times[i] = (u, v, w), where u is the source node, v is the target node,
+      and w is the time it takes for a signal to travel from source to target.
+
+      Now, we send a signal from a certain node K.
+
+      How long will it take for all nodes to receive the signal?
+      If it is impossible, return -1.
+
+      times: List[List[int]], N: int, K: int) -> int:
+
+      Example
+      Input: times = [[2,1,1],[2,3,1],[3,4,1]], N = 4, K = 2
+      Output: 2
+
+      Runtime: 496 ms, faster than 73.14% of Python3 online submissions
+      for Network Delay Time.
+      Memory Usage: 15.8 MB, less than 19.89% of Python3 online submissions
+      for Network Delay Time.
+
+      Overall time complexity...
+      O(E+N+E+N^2) = O(N^2+N+2E)
+      O(E) to build the graph, where E is the number of items in `times`
+      O(N+E) to do breadth first search on the graph to check whether everything
+      is connected
+      O(N^2) to do Dijkstra's algorithm to find the network delay time
+
+      if it is impossible to reach all nodes,
+      running time is O(E+N+E) = O(2E+N)
+    """
+    # if cannnot reach all the nodes, it is impossible, return -1
+    # breadth first search. if not all are connected, impossible. return -1
+
+    from collections import defaultdict, deque
+    # defaultdict so we don't have to handle keyerror
+    # dequeue so queueing and dequeueing is o(1) time
+    graph = defaultdict(list)
+
+    # build graph
+    # time complexity, O(E) where E is the number of items in `times`
+    # each item in `times` is actually an edge, right?
+    for source, target, time in times:
+      graph[source].append((target, time))
+
+    # check whether it is possible to reach all nodes with breadth first search
```

Comment:
Do you need this extra BFS step here (can you use the output of Dijkstra's algorithm to answer this same question)?